File: sort/applied/two_sum_two_pointers/three_sum_closest.scala
object Solution {
    def threeSumClosest(nums: Array[Int], target: Int): Int = {
        // q: greedy or a dynamic programming? dynamic = two sum problem ++
        
        // two sum 
        // 1. brute force w/ for loops
        // 2. map to track location of values 
        // 3. sort the array + two-pointers -> best choice
        
        // only need one solution so don't need to have solutions accumulator etc.
        // scala choice: immutable data structures & recursive function calls or
        // mutable data structures & foldLeft? 
        // if foldLeft need to keep track of two indices and have if statement on leftIndex < rightIndex
        // cleaner to go for recursive functional calls
        
        // not pure dynamic programming = because we need to test each sum (including num) against target
        val sortedNums = nums.sorted
        
        def recursiveThreeSum(numIndex: Int, leftIndex: Int, rightIndex: Int, target: Int, closestDiff: Int): Int = {
            val sum = sortedNums(numIndex) + sortedNums(leftIndex) + sortedNums(rightIndex)
            val diff = target - sum
            val newClosestDiff = if(diff.abs < closestDiff.abs) diff else closestDiff
            if(sum < target && leftIndex + 1 < rightIndex)
                recursiveThreeSum(numIndex, leftIndex + 1, rightIndex, target, newClosestDiff)
            else if(sum > target && leftIndex < rightIndex - 1)
                recursiveThreeSum(numIndex, leftIndex, rightIndex - 1, target, newClosestDiff)
            else if(numIndex + 1 < sortedNums.size - 2) // need room for leftIndex and rightIndex
                recursiveThreeSum(numIndex + 1, numIndex + 2, sortedNums.size - 1, target, newClosestDiff)
            else
                target - newClosestDiff
        }
        recursiveThreeSum(0, 1, sortedNums.size - 1, target, Int.MaxValue)
    }
}

File: data_structures/two_pointers/sorted_values/three_sum_closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        // greedy or dynamic programming? dynamic because can treat as similar to two-sum problem
        // single answer = do not need to accumulate solutions and can return immediately
        // closest to target = either 1. sort the array and choose the closest 2. retain in memory the closest
        
        // two sum problem
        // 1. brute force two for loops
        // 2. hash map
        // 3. sort first then two pointers
        Arrays.sort(nums);
        int closestDifference = Integer.MAX_VALUE;
        for(int i = 0; i < nums.length; i++){
            closestDifference = twoSum(nums, target, i, closestDifference);
        }
        return target - closestDifference;
    }
    
    // key thing is cannot use closestSum because then start with Integer.MAX_VALUE and can get stuck
    // need to use closestDifference
    private int twoSum(int[] nums, int target, int numIndex, int closestDifference){
        // assume nums is sorted
        int leftIndex = numIndex + 1;
        int rightIndex = nums.length - 1;
        while(leftIndex < rightIndex){
            int sum = nums[numIndex] + nums[leftIndex] + nums[rightIndex];
            if(Math.abs(target - sum) < Math.abs(closestDifference)){
                closestDifference = target - sum;
            }
            if(sum == target){
                return 0;
            }
            else if(target - sum > 0){
                leftIndex++;
            }
            else { // target - sum < 0
                rightIndex--;
            }
        }
        return closestDifference;
    }
}